geoh5py/ui_json/input_file.py

```diff
@@ -0,0 +1,577 @@
+from geoh5py.shared.validators import UUIDValidator
+from geoh5py.workspace import Workspace
+
+from .constants import base_validations, ui_validations
+from .validation import InputValidation
+
+
+class InputFile:
+    """
+    Handles loading ui.json input files.
+
+    Attributes
+    ----------
+    data : Input file content parsed to flat dictionary of key:value.
+    ui_json: User interface serializable as ui.json format
+    workspace: Target :obj:`geoh5py.workspace.Workspace`
+    validations: Dictionary of validations for parameters in the input file
+
+    Methods
+    -------
+    write_ui_json()
+        Writes a ui.json formatted file from 'data' attribute contents.
+    read_ui_json()
+        Reads a ui.json formatted file into 'data' attribute dictionary.  Optionally filters
+        ui.json fields other than 'value'.
+    """
+
+    _ui_validators = None
+    _validators = None
+    uuid_validator = UUIDValidator()
+
+    def __init__(
+        self,
+        data: dict[str, Any] = None,
+        ui_json: dict[str, Any] = None,
+        validations: dict = None,
+        validation_options: dict = None,
+        workspace: Workspace = None,
+    ):
+        self._validation_options = validation_options
+        self.validations = validations
+        self.ui_json = ui_json
+        self.data = data
+        self.workspace = workspace
+
+    @property
+    def data(self):
+        if (
+            getattr(self, "_data", None) is None
+            and getattr(self, "_ui_json", None) is not None
+        ):
+            self.data = self.flatten(self.ui_json)
+
+        return self._data
+
+    @data.setter
+    def data(self, value: dict[str, Any]):
+        if value is not None:
+            if not isinstance(value, dict):
+                raise ValueError("Input 'data' must be of type dict or None.")
+
+            if "geoh5" in value:
+                self.workspace = value["geoh5"]
+
+            value = self._promote(value)
+
+            if self.validators is not None:
+                self.validators.validate_data(value)
+
+        self._data = value
+
+    @property
+    def ui_json(self) -> dict | None:
+        """
+        Dictionary representing the ui.json file with promoted values.
+        """
+        return self._ui_json
+
+    @ui_json.setter
+    def ui_json(self, value: dict[str, Any]):
+        if value is not None:
+
+            if not isinstance(value, dict):
+                raise ValueError("Input 'ui_json' must be of type dict or None.")
+
+            self._ui_json = self._numify(value)
+            default_validations = self.get_default_validations(self._ui_json)
+            for key, validations in default_validations.items():
+                if key in self.validations:
+                    validations = {**validations, **self.validations[key]}
+                self._validations[key] = validations
+        else:
+            self._ui_json = None
+        self._validators = None
+
+    @staticmethod
```

Comment:
I think static makes sense since we can instantiate the class directly on read?

Reply:
I think static and class methods should both work more or less the same, but the class method already has the class there to instantiate and return:

```
@classmethod
def read_ui_json(cls, json_file: str):
    input_file = cls()
    blah blah
    return input_file
```
and I still think it should be right after `__init__` since it's a constructor pattern.